coordinator/ethereum/message-anchoring-i2/src/test/kotlin/linea/anchoring/MessageAnchoringAppTest.kt

```diff
@@ -0,0 +1,268 @@
+          .isEqualTo(ethLogs.last().l1RollingHashUpdated.event.rollingHash)
+
+        assertThat(l2MessageService.getAnchoredMessageHashes().map { it.encodeHex() })
+          .isEqualTo(ethLogs.drop(1).map { it.messageSent.event.messageHash.encodeHex() })
+      }
+
+    anchoringApp.stop().get()
+  }
+
+  @Test
+  fun `should anchor messages in chunks when too many messages are waiting on L1`() {
+    val ethLogs = createL1MessageSentV1Logs(
+      l1BlocksWithMessages = (1UL..100UL).map { it },
+      numberOfMessagesPerBlock = 1
+    )
+    val anchoringApp = createApp(
+      l1EventSearchBlockChunk = 10u,
+      maxMessagesToAnchorPerL2Transaction = 20u,
+      anchoringTickInterval = 1.seconds
+    )
+    anchoringApp.start().get()
+
+    addLogsToFakeEthClient(ethLogs)
+    l1Client.setFinalizedBlockTag(ethLogs.last().messageSent.log.blockNumber + 10UL)
+
+    await()
+      .atMost(10.seconds.toJavaDuration())
+      .untilAsserted {
+        assertThat(l2MessageService.getLastAnchoredL1MessageNumber(block = BlockParameter.Tag.LATEST).get())
+          .isEqualTo(ethLogs.last().l1RollingHashUpdated.event.messageNumber)
+      }
+
+    anchoringApp.stop().get()
+  }
+
+  @Test
+  fun `should be resilient to low activity or downtimes and find messages on L1 spread apart and anchor them`() {
+    // put Log every 1000 Blocks on L1
+    // search chunk is 10 so it needs 100 queries to find 1 log on L1
+    // slow timeout 1
+    val ethLogs = createL1MessageSentV1Logs(
+      l1BlocksWithMessages = (1UL..100_000UL).step(1000).map { it },
+      numberOfMessagesPerBlock = 1
+    )
+    val anchoringApp = createApp(
+      l1PollingInterval = 1.milliseconds,
+      l1EventSearchBlockChunk = 10u,
+      l1EventPollingTimeout = 50.milliseconds,
+      l1SuccessBackoffDelay = 20.milliseconds,
+      maxMessagesToAnchorPerL2Transaction = 50u,
+      anchoringTickInterval = 20.milliseconds
+    )
+
+    addLogsToFakeEthClient(ethLogs)
+    l1Client.setFinalizedBlockTag(ethLogs.last().messageSent.log.blockNumber + 10UL)
+
+    anchoringApp.start().get()
+    await()
+      .atMost(10.minutes.toJavaDuration())
+      .untilAsserted {
+        assertThat(l2MessageService.getLastAnchoredL1MessageNumber(block = BlockParameter.Tag.LATEST).get())
+          .isEqualTo(ethLogs.last().l1RollingHashUpdated.event.messageNumber)
+      }
+
+    assertThat(l2MessageService.getAnchoredMessageHashes().map { it.encodeHex() })
+      .isEqualTo(ethLogs.map { it.messageSent.event.messageHash.encodeHex() })
+
+    anchoringApp.stop().get()
+  }
+
+  private fun createL1MessageSentV1Logs(
+    l1BlocksWithMessages: List<ULong>,
+    numberOfMessagesPerBlock: Int,
+    startingMessageNumber: ULong = 1UL
+  ): List<L1MessageSentV1EthLogs> {
+    val ethLogs = mutableListOf<L1MessageSentV1EthLogs>()
+    var messageNumber = startingMessageNumber
+    l1BlocksWithMessages.forEach { blockNumber ->
+      for (i in 0 until numberOfMessagesPerBlock) {
```

Comment:
Minor: i isn't used. I think this can be replaced by `repeat(numberOfMessagesPerBlock)`